File: backend/app/schemas/goals.py
import uuid
from datetime import date, datetime

from pydantic import BaseModel

from app.models.base import ScopeType


class GoalCreate(BaseModel):
    title: str
    description: str | None = None
    scope_type: ScopeType = ScopeType.USER
    scope_id: uuid.UUID | None = None
    success_metric_type: str | None = None
    target_value: float | None = None
    target_date: date | None = None
    priority_weight: float = 1.0
    min_weekly_hours: float | None = None
    max_weekly_hours: float | None = None
    constraints: dict | None = None


class GoalUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    success_metric_type: str | None = None
    target_value: float | None = None
    target_date: date | None = None
    priority_weight: float | None = None
    min_weekly_hours: float | None = None
    max_weekly_hours: float | None = None
    constraints: dict | None = None
    is_active: bool | None = None


class GoalResponse(BaseModel):
    id: uuid.UUID
    title: str
    description: str | None
    scope_type: ScopeType
    scope_id: uuid.UUID
    success_metric_type: str | None
    target_value: float | None
    target_date: date | None
    priority_weight: float
    min_weekly_hours: float | None
    max_weekly_hours: float | None
    constraints: dict | None
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class TaskCreate(BaseModel):
    title: str
    description: str | None = None
    estimated_minutes: int | None = None
    difficulty: int | None = None
    due_date: date | None = None
    dislike_score: int = 0
    owner_user_id: uuid.UUID | None = None
    prerequisites: list | None = None


class TaskUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    estimated_minutes: int | None = None
    difficulty: int | None = None
    due_date: date | None = None
    dislike_score: int | None = None
    owner_user_id: uuid.UUID | None = None
    prerequisites: list | None = None
    status: str | None = None


class TaskResponse(BaseModel):
    id: uuid.UUID
    goal_id: uuid.UUID
    title: str
    description: str | None
    estimated_minutes: int | None
    difficulty: int | None
    due_date: date | None
    dislike_score: int
    owner_user_id: uuid.UUID | None
    prerequisites: list | None
    status: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

File: frontend/lib/api/goals.ts
"use client";

import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import { apiClient } from "@/lib/api/client";
import { toApiError } from "@/lib/api/errors";

/** Mirrors backend `GoalResponse` (app/schemas/goals.py).
 *
 * Dates arrive as strings, not Date objects: `target_date` is a bare
 * "YYYY-MM-DD" (pydantic `date`), while the timestamps are ISO-8601 with a
 * timezone (pydantic `datetime`). Parsing `target_date` with `new Date()`
 * would treat it as UTC midnight and shift the day in negative offsets, so it
 * stays a string all the way to the `<input type="date">`, which expects
 * exactly that format.
 */
export type Goal = {
  id: string;
  title: string;
  description: string | null;
  scope_type: "user" | "team" | "org";
  scope_id: string;
  success_metric_type: string | null;
  target_value: number | null;
  target_date: string | null;
  priority_weight: number;
  min_weekly_hours: number | null;
  max_weekly_hours: number | null;
  constraints: Record<string, unknown> | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
};

/** Body for POST /goals. Backend defaults scope to the current user. */
export type GoalCreateInput = {
  title: string;
  description: string | null;
  success_metric_type: string | null;
  target_value: number | null;
  target_date: string | null;
  priority_weight: number;
  min_weekly_hours: number | null;
  max_weekly_hours: number | null;
};

/** Body for PATCH /goals/{id}.
 *
 * The backend applies `model_dump(exclude_unset=True)`, so an omitted key is
 * left untouched while an explicit `null` clears the column. Keep this a
 * Partial so callers can express that difference.
 */
export type GoalUpdateInput = Partial<GoalCreateInput> & { is_active?: boolean };

export const goalsQueryKey = ["goals"] as const;

export function useGoals() {
  return useQuery({
    queryKey: goalsQueryKey,
    queryFn: async (): Promise<Goal[]> => {
      const { data } = await apiClient.get<Goal[]>("/goals");
      return data;
    },
  });
}

export function useCreateGoal() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async (input: GoalCreateInput): Promise<Goal> => {
      try {
        const { data } = await apiClient.post<Goal>("/goals", input);
        return data;
      } catch (error) {
        throw toApiError(error, "Could not create goal");
      }
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: goalsQueryKey });
    },
  });
}

export function useUpdateGoal() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async ({
      id,
      ...input
    }: GoalUpdateInput & { id: string }): Promise<Goal> => {
      try {
        const { data } = await apiClient.patch<Goal>(`/goals/${id}`, input);
        return data;
      } catch (error) {
        throw toApiError(error, "Could not update goal");
      }
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: goalsQueryKey });
    },
  });
}

export function useDeleteGoal() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async (id: string): Promise<void> => {
      try {
        await apiClient.delete(`/goals/${id}`);
      } catch (error) {
        throw toApiError(error, "Could not delete goal");
      }
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: goalsQueryKey });
    },
  });
}
